tests/test_charm.py

```diff
@@ -82,17 +83,13 @@ def test_alerting_config_is_updated_by_alertmanager_relation(self):
         self.harness.update_relation_data(rel_id,
                                           'alertmanager',
                                           {
-                                              'port':
-                                              '9093'
-                                          })
-        self.harness.update_relation_data(rel_id, 'alertmanager/0',
-                                          {
-                                              'ingress-address':
-                                              '192.169.0.1'
+                                              'port': '9093',
+                                              'addrs': '["192.168.0.1"]'
                                           })
         pod_spec = self.harness.get_pod_spec()
         self.assertEqual(alerting_config(pod_spec), SAMPLE_ALERTING_CONFIG)
 
+    @unittest.skip('This test will not work until operator bug #452 is fixed')
```

Comment:
I think skipping this test is hiding a problem where relation departed hasn't been updated to reflect recent changes. It looks like `alertmanager_relation_departed` would only work on the previous dictionary data structure of `_stored.alertmanagers`.

Reply:
When I commented out the even.unit is None check, both tests passed. Could you be more specific about how this won't work?

note: I do see an issue with on_alertmanager_broken which I will fix

Reply:
@dstathis 

Failure of this test highlights a bug in the charm source code as pointed out by @justinmclark . The bug is in the relation departed event handler. You seem to be suggesting that the best way to fix this bug, is to remove the relation departed event handler entirely. That this is the best way to fix the bug is not entirely clear to me. Prior to your recent changes the relation departed handler was intended to ensure, when a single unit of alertmanager departs the relation, (but others may still be present), then the alerting configuration of the prometheus pod specification does reflect these changes. Is this behaviour still maintained after removing the relation departed handler from Prometheus ? If so please make explicit how. And if so is it possible to modify this test to check for the fact or add a different test to do so ?

Reply:
Apologies for not being more explicit yesterday. The main problem was that that the old structure of `_stored.altermanagers` was a dictionary keyed on `relation_id`, so a `relation_departed` event could remove individual units.

Now that `_stored.alertmanagers` is a list, there doesn't seem to be a good way of removing individual units. Could it go back to being a dict? Or does that mess things up from the alertmanager side of things?

I don't think we can just use `relation_broken`, because that event is fired only once for the whole relation. `relation_departed` (from my understanding) is necessary to keep because if there is ever a scale down, we don't want to remove all information from the whole relation - just individual units that are being removed.

Reply:
Alertmanager provides this list all in one go so just copying the new value from relation data is all that is needed and a change to that data will trigger relation_changed in prometheus. So departed is not needed. The question is whether anything triggers on the alertmanager charm when a unit departs. This is needed so that alertmanager can update it's relation data with the new addresses. It is not clear to me that this happens

Reply:
@dstathis not sure what you mean by "anything triggers on the alertmanager charm when a unit departs". My understanding is Alertmanager charm will get a relation departed event, when an alermanager unit is lost, if the alertmanager charm defines a peer relation.